gpa.py

```diff
@@ -121,24 +141,44 @@ def compute_results(student_records):
    dictionary - cgpa
    """
    #get unique semester names
-   semesters= get_semesters(student_records)
-   total_units, total_grade_points =[],[]
-   results = {}
-   for term in semesters:
-      units, scores =get_score_data(term,student_records )
-      tu,tgps,gpa=compute_gpa(units, scores)
-      total_units.append(tu)
-      total_grade_points.append(tgps)
-      results[term]=gpa
-
-   cgpa = round(sum(total_grade_points) / sum(total_units),2)
-   results['name'] = student_records['name']
-   results['id'] = student_records['id']
-   results['CGPA'] = cgpa
-
-   return results
-
-# print(compute_results(student_records))
+   sessions = get_sessions(student_records)
+   semesters = get_semesters(student_records)
+   session_total_units, session_total_grade_points =[],[]
+
+   final_year_results = {}
+   for session in sessions:
+      total_units, total_grade_points =[],[]
+      results = {}
+      for term in semesters:
+         """
+         compute cgpa for all terms in each session and keeping track of the units $ grades
+         """
+         units, scores = get_score_data(session,term,student_records )
+         tu,tgps,gpa = compute_student_gpa(units, scores)
+         total_units.append(tu)
+         total_grade_points.append(tgps)
+         results[term] = gpa
+         cgpa = round(sum(total_grade_points) / sum(total_units),2)
+         results['cgpa'] = cgpa
+
+      """
+      keeping track of the total unit and grade point outside the term loop to be used for the final year computation
+      """
+      session_total_units.append(total_units)
+      session_total_grade_points.append(total_grade_points)
+      
+      """
+      final year computaion of cgpa with values of each session
+      """
+      final_year_cgpa = round(sum(calculate_sum(session_total_grade_points)) / sum(calculate_sum(session_total_units)),2)
+      final_year_results['name'] = student_records['name']
+      final_year_results['id'] = student_records['id']
+      final_year_results[f"{session}_session"] = results
+      final_year_results['final_year_cgpa'] = final_year_cgpa
```

Comment:
Why is this nested in a for loop?

Reply:
And by the way, I'm not sure the current solution will be able to report details such as: 
Year: 2019, Level: 100L
Semester: Fall
     total unit, GPA, CGPA
Semester: Spring
     total unit, GPA, CGPA
Year: 2020, Level: 200L
Semester: Fall
     total unit, GPA, CGPA
Semester: Spring
     total unit, GPA, CGPA
.....
.....
Year: 2022, Level: 400L
Semester: Fall
     total unit, GPA, CGPA
Semester: Spring
     total unit, GPA, CGPA
.....
total units, CPGA

Reply:
I moved the final computation outside the nested loop.
Refactored the code. Used a different data structure(Tuple) as the dictionary gave some issues around duplicate keys during the loop.